### PR TITLE
Add x and r emitters

### PR DESCRIPTION
**Pull request by roelof-groenewald**
`XPlaneEmitter` and `ZCylinderEmitter` were added to emit from the side walls of either `XZ` or `RZ` simulations. The `XPlaneEmitter` follows easily from the `ZPlaneEmitter` so it was not heavily tested. The `ZCylinderEmitter` however was compared to the known vacuum diode solution to ensure that the velocity distribution of emitted particles were properly sampled. This also served as a test of the `rmin != 0` functionality in `RZ` simulations in WarpX.
A 50 um gap was simulated with `r_max = 1e-3` m, the cathode was on the outer wall of the tube. The area of the anode in this case is about 5% smaller than the cathode's and therefore we should expect only small deviations from theory due to the `RZ` nature of the problem. The input file used for the data points below was also added as an example file.

D_50

Still to do:
- [x] test the EB functionality for `InfCylinderZ` - a test was added to check a point on the vacuum diode IV curve for accuracy
- [x] update changelog and version number

The changes to `ElectrostaticSolver.cpp` now allows us to actually use a Nuemann boundary condition for the `r_max` boundary. Due to this change I reran the vacuum diode simulation with the `ZDiscEmitter` as was done in PR #121. The results are shown below and still agrees well with theory.

D_50

### CODE REVIEW

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+mwxutil.init_libwarpx(ndim=2, rz=True)
+
+import numpy as np
+from pywarpx import picmi
+
+from mewarpx import assemblies, diags, emission, mespecies, runinfo
+from mewarpx.mwxrun import mwxrun
+
+
+class VacuumTEC_Cylinder(object):
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py]:
```suggestion
class CylinderVacuumTEC(object):
```
Underscores in python classes aren't standard, so if you don't like `TECCylinder` this seems like a good alternative.

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+parser = argparse.ArgumentParser()
+parser.add_argument('--V', help='bias voltage in Volt', type=float, default=0)
+parser.add_argument('--steps', help='set the number of simulation steps',
+                    type=int, default=3000)
+parser.add_argument('--save', help='save voltage and current pairs',
+                    default=False, action='store_true')
+args, left = parser.parse_known_args()
+sys.argv = sys.argv[:1]+left
+
+run = VacuumTEC_Cylinder()
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py]:
```suggestion
run = CylinderVacuumTEC(
    V_ANODE_CATHODE=args.V, TOTAL_TIMESTEPS=args.steps, SAVE=args.save
)
```

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+run.V_ANODE_CATHODE = args.V
+run.TOTAL_TIMESTEPS = args.steps
+run.SAVE = args.save
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py]:
```suggestion
```
Passing these as init arguments is a much cleaner separation between class interface and outside code.

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    D_CA = 50e-6                 # m
+
+    NZ = 32
+    NR = 512
+
+    DT = 0.5e-12                 # s
+
+    NPPC = 4
+
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py:38]:
```suggestion

    def __init__(self, V_ANODE_CATHODE, TOTAL_TIMESTEPS, SAVE):
        self.V_ANODE_CATHODE = V_ANODE_CATHODE
        self.TOTAL_TIMESTEPS = TOTAL_TIMESTEPS
        self.SAVE = SAVE
        
```

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    ANODE_TEMP = 200             # K
+    ANODE_PHI = 1.4              # eV
+
+    D_CA = 50e-6                 # m
+
+    NZ = 32
+    NR = 512
+
+    DT = 0.5e-12                 # s
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py:35]: What voltage does this satisfy CFL up to? This seems like a place a user could easily confuse themselves with bad results.

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            lower_bound=[self.rmin, self.zmin],
+            upper_bound=[self.rmax, self.zmax],
+            number_of_cells=[self.NR, self.NZ], min_tiles=1,
+            bc_fields_z_min='periodic', bc_fields_z_max='periodic',
+            bc_particles_z_min='periodic', bc_particles_z_max='periodic',
+            bc_fields_r_min='dirichlet', bc_fields_r_max='dirichlet',
+        )
+        mwxrun.init_timestep(DT=self.DT)
+        mwxrun.simulation.max_steps = self.TOTAL_TIMESTEPS
+        mwxrun.simulation.load_balance_intervals = self.TOTAL_TIMESTEPS // 5000
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py:80]: What happens if this evaluates to 0? Does that just disable load balancing? (I honestly don't know, just want to make sure it's not an error to set it to 0.)

File: mewarpx/examples/thermionic_diode_rz_cylinder.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            emitter=self.cathode_emitter, species=self.electrons,
+            A=self.CATHODE_A, npart_per_cellstep=self.NPPC,
+            use_Schottky=self.USE_SCHOTTKY
+        )
+
+        #######################################################################
+        # Clean up and output RunInfo                                         #
+        #######################################################################
+
+        runvars = VacuumTEC_Cylinder.__dict__.copy()
```
Comment (peterscherpelz) [inline mewarpx/examples/thermionic_diode_rz_cylinder.py]:
```suggestion
        runvars = CylinderVacuumTEC.__dict__.copy()
```

File: mewarpx/mewarpx/assemblies.py [hunk 8]
```diff
@@ -371,8 +387,114 @@ def calculatenormal(self, px, py, pz):
...
+        if self.r_inner == mwxrun.rmax:
+            # set solver boundary potential
+            mwxrun.grid.potential_xmax = self.V
+            # set label to correctly grab scraped particles from the buffer
+            self.scraper_label = 'x_hi'
+        elif self.r_outer == mwxrun.rmin:
+            # set solver boundary potential
+            mwxrun.grid.potential_xmin = self.V
+            # set label to correctly grab scraped particles from the buffer
+            self.scraper_label = 'x_lo'
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/assemblies.py]: I would probably use `np.isclose` with very low `atol` and `rtol`; a script that accidentally computes `rmax` or `rmin` twice with different arithmetic orders could lead to a failure here.

File: mewarpx/mewarpx/assemblies.py [hunk 9]
```diff
@@ -371,8 +387,114 @@ def calculatenormal(self, px, py, pz):
...
+        else:
+            # negative means outside of object
+            self.implicit_function = (
+                f"-(abs(x-{self.r_center})-{self.r_outer - self.r_center})"
+            )
+            # set label to correctly grab scraped particles from the buffer
+            self.scraper_label = 'eb'
+            # install the EB in the simulation
+            self._install_in_simulation()
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/assemblies.py]: I don't get this use case and I don't see it being tested anywhere. I can see a use-case in an XYZ simulation, but we don't support that yet, and even then we'd probably want a finite cylinder and/or the ability to set x and y to nonzero values. In the RZ case, this embeds a boundary that spans the whole domain, which I don't think we'd ever want.

Maybe comment this out, and raise a NotImplementedError that says it should only be for XYZ geometry and such use is untested?

File: mewarpx/mewarpx/assemblies.py [hunk 10]
```diff
@@ -371,8 +387,114 @@ def calculatenormal(self, px, py, pz):
...
+            nhat (np.ndarray): Normal of particles of conductor to nearest
+                surface of conductor.
+        """
+        # distance from center of cylinder
+        dist = np.sqrt((px - self.center_x)**2 + (py - self.center_y)**2)
+        nhat = np.zeros([3, len(px)])
+        nhat[0, :] = (px - self.center_x) / dist
+        nhat[2, :] = (pz - self.center_z) / dist
+
+        # flip the normal vector for points inside the cylinder
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/assemblies.py]:
```suggestion
        # Flip the normal vector for points inside the cylinder inner wall,
        # or inside the cylinder assembly but closer to the outer wall than the inner.
```
it looks like is what the below statement is doing?

File: mewarpx/mewarpx/emission.py [hunk 11]
```diff
@@ -1179,20 +1180,155 @@ def get_normals(self, x, y, z):
...
+        if self.xsign not in [-1, 1]:
+            raise ValueError("xsign must be +1 or -1 for x-plane emitters.")
+
+        self.x = x
+        if self.x is None:
+            try:
+                self.x = conductor.x
+            except AttributeError:
+                try:
+                    attr = 'xmax' if self.xsign else 'xmin'
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: `bool(1) == bool(-1) == True`, so I don't think this logic works?

File: mewarpx/mewarpx/emission.py [hunk 12]
```diff
@@ -1179,20 +1180,155 @@ def get_normals(self, x, y, z):
...
+        self.bounds = []
+        bounds_tuples = []
+        for bound in ['ymin', 'ymax', 'zmin', 'zmax']:
+            default = (
+                getattr(mwxrun, bound) if not hasattr(conductor, bound) else
+                getattr(conductor, bound)
+            )
+            bounds_tuples.append((locals()[bound], default))
+        for coord, default in bounds_tuples:
+            self.bounds.append(coord if coord is not None else default)
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]:
```suggestion
            bound = locals()[bound]
            if bound is None:
                bound = default
            self.bounds.append(bound)
```
or `self.bounds.append(locals()[bound] if localcs()[bound] is not None else default)` if you want a one-liner I believe?

Anyway, both of those seem more immediately interpretable than the extra for loop to me...

File: mewarpx/mewarpx/emission.py [hunk 13]
```diff
@@ -1179,20 +1180,155 @@ def get_normals(self, x, y, z):
...
+                except AttributeError:
+                    raise AttributeError(
+                        'x must be specified for x-plane emitter if the '
+                        f'attached conductor does not specify x or {attr}'
+                    )
+
+        # Determine bounds
+        # Will be 4 element array [ymin, ymax, zmin, zmax]
+        self.bounds = []
+        bounds_tuples = []
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]:
```suggestion
```
If you use below suggestion too.

File: mewarpx/mewarpx/emission.py [hunk 14]
```diff
@@ -1277,6 +1414,148 @@ def get_normals(self, x, y, z):
...
+        if self.rsign not in [-1, 1]:
+            raise ValueError("rsign must be +1 or -1 for z-cylinder emitters.")
+        if self.rsign == 1:
+            self.r = conductor.r_outer + 1e-10
+        else:
+            self.r = conductor.r_inner - 1e-10
+        self.transverse_fac = transverse_fac
+
+        # sanity check
+        if self.r == 0:
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]:
```suggestion
        if self.r <= 0:
```
Otherwise the `- 1e-10` just above will cause this check to be skipped for `r_inner = 0`.

File: mewarpx/mewarpx/emission.py [hunk 15]
```diff
@@ -1277,6 +1414,148 @@ def get_normals(self, x, y, z):
...
+        normals[z < self.zmin, 2] = -1.0
+
+        # On the side face, the normal is r-hat
+        normal_idxs = (z < self.zmax) & (z > self.zmin)
+        x_side = x[normal_idxs]
+        y_side = y[normal_idxs]
+        r_side = np.sqrt(x_side**2 + y_side**2)
+        normals[normal_idxs, 0] = x_side / r_side
+        normals[normal_idxs, 1] = y_side / r_side
+        return normals
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: I would keep only the side face stuff here. That's because these are the normals for emitted electrons, and this class only emits from the side. Whereas you're porting from one that can actually emit from top & bottom faces.
```suggestion
        # The normal is r-hat
        r = np.sqrt(x**2 + y**2)
        normals[:, 0] = x / r
        normals[:, 1] = y / r
        return normals
```

**Review by peterscherpelz — CHANGES_REQUESTED** [review]
I would recommend changing `xsign` and `rsign` to `xdir` and `rdir`. This is because these signs are effectively opposite the convention used for `zsign`, which due to implementation of semi-infinite planes are -1 for +z emission and 1 for -z emission. By using a different variable name, I think it's less likely anyone will confuse signs going from one to the other.

Thanks for implementing these! Let me know if any of my suggestions don't make sense, that was a lot to go through...

File: mewarpx/examples/thermionic_diode_rz_cylinder.py (hunk 6 above)
Comment (roelof-groenewald) [reply mewarpx/examples/thermionic_diode_rz_cylinder.py:80]: Yeap, the default value is 0 so in that case load balancing just doesn't occur.

File: mewarpx/mewarpx/assemblies.py (hunk 9 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/assemblies.py]: This is now tested and it works. It does the same thing as having `r_min != 0`, so I agree that we'll probably not use it until we do 3D sims but seeing as it tests another part of the code - scraping from EBs in RZ - I vote to keep it.

File: mewarpx/mewarpx/emission.py (hunk 11 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py]: Thanks!

File: mewarpx/examples/thermionic_diode_rz_cylinder.py (hunk 5 above)
Comment (roelof-groenewald) [reply mewarpx/examples/thermionic_diode_rz_cylinder.py:35]: Not sure, but it looks like it is properly resolved up to 8 V where the full J_sat is already pulled out. So if someone changes the cathode temperature or WF they should also confirm that the simulation is still resolved (also with regards to the spatial resolution).

File: mewarpx/changelog.csv [hunk 16]
```diff
@@ -1,5 +1,34 @@
...
+- The electrostatic solver was updated to allow a Dirichlet BC at ``rmin`` for
+  RZ simulations if ``rmin != 0``.
+
+**Bugfixes**:
+
+- The default boundary condition at ``rmax`` for RZ simulations was set to be
+  Neumann, but the actual BC was hard coded to be Dirichlet with 0 V potential.
+  This has now been fixed.
+"
+4.0.0, 2, 12/31/2021, "
```
Comment (roelof-groenewald) [inline mewarpx/changelog.csv]: Put in appropriate date when ready to merge into `memaster`.

File: mewarpx/mewarpx/assemblies.py (hunk 9 above)
Comment (peterscherpelz) [reply mewarpx/mewarpx/assemblies.py]: ah, testing EBs in RZ is a great point, yeah

File: mewarpx/mewarpx/emission.py [hunk 17]
```diff
@@ -1179,20 +1180,158 @@ def get_normals(self, x, y, z):
...
+        self.bounds = []
+        for bound in ['ymin', 'ymax', 'zmin', 'zmax']:
+            default = (
+                getattr(mwxrun, bound) if not hasattr(conductor, bound) else
+                getattr(conductor, bound)
+            )
+            bound = locals()[bound]
+            if bound is None:
+                bound = default
+            self.bounds.append(bound)
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: Gah, even though I technically overwrote bound after it was done being used I didn't mean to. How about
```suggestion
        for boundstr in ['ymin', 'ymax', 'zmin', 'zmax']:
            bound = locals()[boundstr]
            if bound is None:
                bound = (
                    getattr(mwxrun, boundstr) if not hasattr(conductor, boundstr) else
                    getattr(conductor, boundstr)
                )
            self.bounds.append(bound)
```

**Review by peterscherpelz — APPROVED** [review]
One more suggestion otherwise it looks good; you can update changelog dates/git hash and merge if you'd like.

File: mewarpx/changelog.csv [hunk 18]
```diff
@@ -28,7 +28,7 @@ Version, Physics version, Date,        List of changes
   Neumann, but the actual BC was hard coded to be Dirichlet with 0 V potential.
   This has now been fixed.
 "
-4.0.0, 2, 12/31/2021, "
+4.0.0, 2, 12/29/2021, "
```
Comment (peterscherpelz) [inline mewarpx/changelog.csv]:
```suggestion
4.0.0, 2, 12/23/2021, "
```
This is the 4.0.0 date

Comment (peterscherpelz) [reply mewarpx/changelog.csv]: 12/29/21 should go in line 2 of the file.

File: mewarpx/mewarpx/assemblies.py [hunk 19]
```diff
@@ -417,20 +434,33 @@ def __init__(self, V, T, WF, name, r_outer, r_inner=0.0):
 
         # check if the cylinder forms the simulation boundary
         if np.isclose(self.r_inner, mwxrun.rmax):
+            if not infinite_cyl:
+                raise AttributeError("CylinderZ outside simulation domain.")
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/assemblies.py]:
```suggestion
                raise AttributeError(
                    "CylinderZ on the simulation boundary must span the full "
                    "z domain."
                )
```
You can change the wording if you want - but notice your error text doesn't make a ton of sense in that if r_inner > mwxrun.rmax this error won't appear, and the error is not really that the r bound is necessarily wrong but rather that zmin/zmax are wrong for a cylinder on the boundary.

File: mewarpx/mewarpx/assemblies.py [hunk 20]
```diff
@@ -417,20 +434,33 @@ def __init__(self, V, T, WF, name, r_outer, r_inner=0.0):
...
             # set solver boundary potential
             mwxrun.grid.potential_xmax = self.V
             # set label to correctly grab scraped particles from the buffer
             self.scraper_label = 'x_hi'
         elif np.isclose(self.r_outer, mwxrun.rmin):
+            if not infinite_cyl:
+                raise AttributeError("CylinderZ outside simulation domain.")
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/assemblies.py]:
```suggestion
                raise AttributeError(
                    "CylinderZ on the simulation boundary must span the full "
                    "z domain."
                )
```

**Review by peterscherpelz — COMMENTED** [review]
Alright, sorry, I'm just full of suggestions today :)

File: mewarpx/changelog.csv (hunk 18 above)
Comment (roelof-groenewald) [reply mewarpx/changelog.csv]: I've definitely been updating these incorrectly the whole time. Every time I update the changelog the current version (5.0.0 in this case) remains `In progress` until there is a newer version ready for release at which point the (end)date for the current version is updated. Point being I've been treating it as a deprecation date not a release date.
Happy to switch it but just know that most of the old dates are wrong then.

Comment (peterscherpelz) [reply mewarpx/changelog.csv]: Ah, ok. I see why my previous practice confused you: Because I never knew when PR reviews were going to finish, I just typically left the newest version as "In Progress" always so I didn't have to remember to update it right before the PR merge. But then next time I altered the changelog, I would go to git history and figure out what the last version release date should have been and update it accordingly.

Let's use release date, since that's obviously more standard for newcomers. Since we have git history anyway, I don't care about updating previous ones though.
